File: go/kernel/kernel_api_test.go
// kernel_api_test.go
package main

import (
	"fmt"
	"os"
	"runtime"
	"syscall"
)

func main() {
	fmt.Println("GOOS/GOARCH:", runtime.GOOS, runtime.GOARCH)

	// 1) 커널 정보: uname (syscall.Uname)
	var u syscall.Utsname
	if err := syscall.Uname(&u); err != nil {
		fmt.Println("Uname error:", err)
		os.Exit(1)
	}
	fmt.Println("uname.sysname:", charsToString(u.Sysname[:]))
	fmt.Println("uname.release:", charsToString(u.Release[:]))

	// 2) PID (syscall)
	fmt.Println("getpid:", os.Getpid())

	// 3) 실패 예시: 없는 PID에 signal 0
	err := syscall.Kill(999999, 0)
	if err != nil {
		fmt.Println("kill(999999,0) failed as expected:", err)
	}
}

func charsToString(ca []int8) string {
	b := make([]byte, 0, len(ca))
	for _, c := range ca {
		if c == 0 {
			break
		}
		b = append(b, byte(c))
	}
	return string(b)
}

File: c/kernel/kernel_api_test.c
// kernel_api_test.c
#define _GNU_SOURCE
#include <errno.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <sys/utsname.h>
#include <sys/syscall.h>

int main(void) {
    struct utsname u;
    if (uname(&u) != 0) {
        perror("uname");
        return 1;
    }

    pid_t pid = getpid();                 // glibc wrapper -> syscall
    long tid = syscall(SYS_gettid);       // direct syscall

    printf("uname.sysname : %s\n", u.sysname);
    printf("uname.release : %s\n", u.release);
    printf("getpid        : %d\n", pid);
    printf("gettid(syscall): %ld\n", tid);

    // 실패 케이스로 errno 보기: 존재하지 않는 PID에 signal 보내보기
    int r = syscall(SYS_kill, 999999, 0);
    if (r != 0) {
        printf("kill(999999,0) failed as expected: errno=%d (%s)\n",
               errno, strerror(errno));
    }
    return 0;
}
